includes/class-wc-install.php

```diff
@@ -744,6 +744,18 @@ private static function get_schema() {
   PRIMARY KEY (download_log_id),
   KEY permission_id (permission_id),
   KEY timestamp (timestamp)
+) $collate;
+CREATE TABLE {$wpdb->prefix}wc_order_stats (
+  start_time timestamp DEFAULT '0000-00-00 00:00:00' NOT NULL,
+  num_orders int(11) UNSIGNED DEFAULT 0 NOT NULL,
+  num_items_sold int(11) UNSIGNED DEFAULT 0 NOT NULL,
+  orders_gross_total double DEFAULT 0 NOT NULL,
+  orders_coupon_total double DEFAULT 0 NOT NULL,
+  orders_refund_total double DEFAULT 0 NOT NULL,
+  orders_tax_total double DEFAULT 0 NOT NULL,
+  orders_shipping_total double DEFAULT 0 NOT NULL,
+  orders_net_total double DEFAULT 0 NOT NULL,
+  PRIMARY KEY (start_time)
```

Comment:
@rodrigoprimo and/or @kloon thoughts on how you think the following table would work instead of the above ^ table on a site with a DB the size of woocommerce.com's or bigger?

`order_id | date_created | num_items_sold | gross_total | coupon_total | refund_total | tax_total | shipping_total | net_total`

Then the stats calculations would happen with SQL on the fly, for example: 

`SELECT SUM(num_items_sold) AS num_items_sold FROM wp_wc_order_stats WHERE date_created > xxxx-xx-xx AND date_created < xxxx-xx-xx GROUP BY HOUR(date_created);`

This new proposed table would be much easier to keep accurately synchronized, but I don't want to run into scalability issues.

Reply:
It would be slower to get information out of the new table when compared to the old one. To know how slow, I did a quick test using a test database containing 1.9 million orders. I first created the new table with a `date_created` index and populated it using real order IDs and order dates but randomly generated values for the other fields:

```
CREATE TABLE `wp_wc_order_stats` (
  `order_id` bigint(20) unsigned NOT NULL,
  `date_created` datetime NOT NULL,
  `num_items_sold` int(11) unsigned NOT NULL DEFAULT '0',
  `gross_total` double NOT NULL DEFAULT '0',
  `coupon_total` double NOT NULL DEFAULT '0',
  `refund_total` double NOT NULL DEFAULT '0',
  `tax_total` double NOT NULL DEFAULT '0',
  `shipping_total` double NOT NULL DEFAULT '0',
  `net_total` double NOT NULL DEFAULT '0',
  PRIMARY KEY (`order_id`),
  KEY `date_created` (`date_created`),
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

INSERT INTO wp_wc_order_stats (order_id, date_created) SELECT ID, post_date FROM wp_posts WHERE post_type = 'shop_order';

UPDATE wp_wc_order_stats SET num_items_sold = CAST(RAND()*(100-1)+1 AS UNSIGNED), gross_total = CAST(RAND()*(100-1)+1 AS UNSIGNED), coupon_total = CAST(RAND()*(100-1)+1 AS UNSIGNED), refund_total = CAST(RAND()*(100-1)+1 AS UNSIGNED), tax_total = CAST(RAND()*(100-1)+1 AS UNSIGNED), shipping_total = CAST(RAND()*(100-1)+1 AS UNSIGNED), net_total = CAST(RAND()*(100-1)+1 AS UNSIGNED);
```

Querying data for short periods of time is relatively fast:

```
SELECT SUM(num_items_sold) AS num_items_sold FROM wp_wc_order_stats WHERE date_created > '2017-12-01' AND date_created < '2017-12-07' GROUP BY HOUR(date_created);
24 rows in set (0.08 sec)
```

It gets slower for longer periods of time:

```
mysql> SELECT SUM(num_items_sold) AS num_items_sold FROM wp_wc_order_stats WHERE date_created > '2016-12-01' AND date_created < '2017-12-01' GROUP BY HOUR(date_created);
24 rows in set (0.69 sec)
```

Here is the explanation of both queries:

```
mysql> explain SELECT SUM(num_items_sold) AS num_items_sold FROM wp_wc_order_stats WHERE date_created > '2017-12-01' AND date_created < '2017-12-07' GROUP BY HOUR(date_created);
+----+-------------+-------------------+------------+-------+---------------+--------------+---------+------+------+----------+--------------------------------------------------------+
| id | select_type | table             | partitions | type  | possible_keys | key          | key_len | ref  | rows | filtered | Extra                                                  |
+----+-------------+-------------------+------------+-------+---------------+--------------+---------+------+------+----------+--------------------------------------------------------+
|  1 | SIMPLE      | wp_wc_order_stats | NULL       | range | date_created  | date_created | 5       | NULL | 7060 |   100.00 | Using index condition; Using temporary; Using filesort |
+----+-------------+-------------------+------------+-------+---------------+--------------+---------+------+------+----------+--------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> explain SELECT SUM(num_items_sold) AS num_items_sold FROM wp_wc_order_stats WHERE date_created > '2016-12-01' AND date_created < '2017-12-01' GROUP BY HOUR(date_created);
+----+-------------+-------------------+------------+------+---------------+------+---------+------+---------+----------+----------------------------------------------+
| id | select_type | table             | partitions | type | possible_keys | key  | key_len | ref  | rows    | filtered | Extra                                        |
+----+-------------+-------------------+------------+------+---------------+------+---------+------+---------+----------+----------------------------------------------+
|  1 | SIMPLE      | wp_wc_order_stats | NULL       | ALL  | date_created  | NULL | NULL    | NULL | 1902250 |    50.00 | Using where; Using temporary; Using filesort |
+----+-------------+-------------------+------------+------+---------------+------+---------+------+---------+----------+----------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

I don't know why MySQL can use an index condition for the first query but cannot to for the second query.

To decide if it is ok to use the new table, I would suggest we first check if there are queries that are potently slower than the one you gave as an example and how fast the queries need to run.

Reply:
Instead of datetime why not use timestamp instead? It uses half the bytes, lighter on space and easier to index. Also a timestamp might be more appropriate for using with timezone specific data.

Reply:
On the flip side, WordPress uses datetime instead of timestamp, so using timestamp might introduce more complexity as developers will need to remember to use two different formats to store dates. I'm not sure which option I prefer, just mentioning one potential downside of timestamp for us to consider.